[Observable] Use AbortSignal directly

maraisr opened the pull request:
#### 39636bf8cda7370d8b98b48ce85bd37263852748
<pre>
[Observable] Use AbortSignal directly
<a href="https://bugs.webkit.org/show_bug.cgi?id=282069">https://bugs.webkit.org/show_bug.cgi?id=282069</a>

Reviewed by Chris Dumez.

As AbortController wraps an AbortSignal, there is no need for our
internal code to go through it as well. We save allocations, and still
entirely spec complaint. Seeing as the spec reads &quot;to signal abort&quot; in various
parts, which is only a concept on abort signals anyway.

* Source/WebCore/dom/InternalObserverFirst.cpp:
(WebCore::createInternalObserverOperatorFirst):
  The visitor signal should just be an AbortSignal, than through the
  AbortController bindings.
  Drive-by: Repeated Ref { m_promise } -&gt; protectedPromise
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::Subscriber):
  Construct a new AbortSignal, than AbortController
(WebCore::Subscriber::close):
  We need to signalAbort the AbortSignal, than abort the AbortController
* Source/WebCore/dom/Subscriber.h:

Canonical link: <a href="https://commits.webkit.org/285723@main">https://commits.webkit.org/285723@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad62a5a593de4f8545531656f694616658739c0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73610 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53039 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77904 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24848 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75725 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62172 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/824 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57869 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16280 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76677 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48030 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63334 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38275 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44618 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20819 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23181 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66356 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21168 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79498 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/927 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/402 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66240 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1069 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63343 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65521 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16195 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9380 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7561 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/891 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/920 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/907 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/926 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

maraisr commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -50,7 +50,9 @@ class Subscriber final : public ActiveDOMObject, public ScriptWrappable, public
 
     bool active() { return m_active; }
-    AbortSignal& signal() { return m_abortController->signal(); }
+    AbortSignal& signal() { return m_signal.get(); }
```
This is referenced by the IDL. So we need a method called `signal`, so things bind.

But question (cc @darinadler @cdumez @rreno):

Is it possible through the IDL bindings, to alias the `signal` operation identifier to an implemented method of `protectedSignal` — and only have one method definition here, and that is protected.

as in:

```diff
- AbortSignal& signal() { return m_signal.get(); }
+ Ref<AbortSignal> protectedSignal() { return m_signal; }
```

```diff
interface Subscriber {
-   readonly attribute AbortSignal signal;
+   [ImplementedAs="protectedSignal"] readonly attribute AbortSignal signal;
//   Just a guess at some syntax?
}
```

Or would this be totally fine to do:

```cpp
// IDL binding
Ref<AbortSignal> signal() const { return m_signal; }
```

and despite convention to use `protected*`, to simply just use `signal()` where we need it.

cdumez replied:
> Ref<AbortSignal> signal() const { return m_signal; }

would not be fine as it doesn't follow our pattern. We usually have both `signal()` which returns a raw pointer/reference and `protectedSignal()` which returns a Ref/RefPtr. It is up to the call site to decide which one to call based on their use case. If the call site only need to call a trivial getter on the signal, then it doesn't need to call `protectedSignal()`, `signal()` will suffice. I do not believe the generated JS bindings NEED a smart pointer here for safety (or the cpp bot would complain).

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -43,11 +42,11 @@ Ref<Subscriber> Subscriber::create(ScriptExecutionContext& context, Ref<Internal
     , m_options(options)
 {
-    followSignal(m_abortController->signal());
+    followSignal(m_signal);
```
Should be `protectedSignal()` here since we're passing as parameter to a non-trivial function.

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -111,13 +110,13 @@ void Subscriber::followSignal(AbortSignal& signal)
 void Subscriber::close(JSC::JSValue reason)
 {
-    auto* context = scriptExecutionContext();
+    RefPtr context = protectedScriptExecutionContext();
```
No need for this change, we can just use `!scriptExecutionContext()` on the next line and avoid the local variable since it is only used in one place.

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -133,17 +132,13 @@ void Subscriber::close(JSC::JSValue reason)
 bool Subscriber::isInactiveDocument() const
 {
-    RefPtr document = dynamicDowncast<Document>(scriptExecutionContext());
+    RefPtr document = dynamicDowncast<Document>(protectedScriptExecutionContext());
```
Please revert this change, there is no benefit to calling `protectedScriptExecutionContext()` here and it makes the code less concise for no good reason.

cdumez commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -133,17 +132,13 @@ void Subscriber::close(JSC::JSValue reason)
 void Subscriber::reportErrorObject(JSC::JSValue value)
 {
-    auto* context = scriptExecutionContext();
-    if (!context)
```
Why did we drop the null check for the context here?

maraisr replied:
No reason, ill revert. Just tried to keep this consistent to the other PR I've been submitting.

maraisr committed:
[Observable] Use AbortSignal directly
https://bugs.webkit.org/show_bug.cgi?id=282069

Reviewed by Chris Dumez.

As AbortController wraps an AbortSignal, there is no need for our
internal code to go through it as well. We save allocations, and still
entirely spec complaint. Seeing as the spec reads "to signal abort" in various
parts, which is only a concept on abort signals anyway.

* Source/WebCore/dom/InternalObserverFirst.cpp:
(WebCore::createInternalObserverOperatorFirst):
  The visitor signal should just be an AbortSignal, than through the
  AbortController bindings.
  Drive-by: Repeated Ref { m_promise } -> protectedPromise
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::Subscriber):
  Construct a new AbortSignal, than AbortController
(WebCore::Subscriber::close):
  We need to signalAbort the AbortSignal, than abort the AbortController
* Source/WebCore/dom/Subscriber.h:

Canonical link: https://commits.webkit.org/285723@main
